### PR TITLE
[wasm] Misc debugger improvements

### PR DESCRIPTION
**Pull request by radical**
- Extract browser provisioning from debugger tests, and make it usable more generally
- Other improvements to the debug proxy, making it more usable from other projects (like the upcoming wasm-app-host)

Extracted from https://github.com/dotnet/runtime/pull/68696

Added by @thaystg:
- `--wait-for-debugger` - If requested, then it will cause invocation of `main` to be delayed till a debugger is attached.
- Add support for setting an automatic breakpoint on the first line of the entrypoint method (`Main`). This will be
useful for debugging with the upcoming wasm-app-host, along with the use of `wait-for-debugger`.

### CODE REVIEW

**Comment by radical** [comment]
cc @radekdoulik I added some minimal `sourceName` changes to the templates.

**Comment by radical** [comment]
wasm/AOT failure is NRE in linker: https://github.com/dotnet/runtime/issues/69104

**Review by thaystg — APPROVED** [review]
LGTM :)

File: src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs [hunk 1]
```diff
@@ -1436,10 +1439,37 @@ await foreach (SourceFile source in context.store.Load(sessionId, loaded_files,
...
                         await OnSourceFileAdded(sessionId, source, context, token);
                     }
+
+                    if (AutoSetBreakpointOnEntryPoint)
+                    {
+                        var entryPoint = context.store.FindEntryPoint();
+                        if (entryPoint is not null)
+                        {
+                            // FIXME: id
+                            string bpId = "bp://99999";
```
Comment (lewing) [inline src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:1449]: do we do anything specific with the bp id format?  If we're making one up can't we do something a little more obvious?

Comment (lewing) [inline src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:1449]: do we do anything specific with the bp id format?  If we're making one up can't we do something a little more obvious?

File: src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs [hunk 2]
```diff
@@ -1350,7 +1345,13 @@ public async IAsyncEnumerable<SourceFile> Load(SessionId id, string[] loaded_fil
         public SourceFile GetFileById(SourceId id) => AllSources().SingleOrDefault(f => f.SourceId.Equals(id));
 
         public AssemblyInfo GetAssemblyByName(string name) => assemblies.FirstOrDefault(a => a.Name.Equals(name, StringComparison.InvariantCultureIgnoreCase));
+        public MethodInfo FindEntryPoint()
+        {
+            if (_entryPoint is null)
+                _entryPoint = assemblies.Where(asm => asm.EntryPoint is not null).Select(asm => asm.EntryPoint).FirstOrDefault();
```
Comment (lewing) [inline src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:1351]: is assemblies guaranteed to be in the correct order for this to work?

Comment (lewing) [inline src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:1351]: is assemblies guaranteed to be in the correct order for this to work?

**Review by lewing — APPROVED** [review]

**Comment by radical** [comment]
Chrome debugger tests are passing.

**Pull request merged**